File: problem-solving/_lc-516-medium-longest-palindromic-subsequence/1-top-down.py
"""
https://leetcode.com/problems/longest-palindromic-subsequence/

Complexity
----------
Time: O(n^2)
Space: O(n^2)
"""

class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        cache = {}
        def lps(l, r):
            if l > r:
                return 0
            if l == r:
                return 1
            if (l, r) in cache:
                return cache[(l, r)]
            if s[l] == s[r]:
                cache[(l, r)] = 2 + lps(l + 1, r - 1)
            else:
                cache[(l, r)] = max(lps(l + 1, r), lps(l, r - 1))
            return cache[(l, r)]
        return lps(0, len(s) - 1)

File: problem-solving/lc-516-medium-longest-palindromic-subsequence/1-top-down.go
/*
https://leetcode.com/problems/longest-palindromic-subsequence/

Complexity
----------
Time: O(n^2)
Space: O(n^2)
*/

func longestPalindromeSubseq(s string) int {
    cache := make(map[string]int)
    var F func(int, int) int
    F = func(l, r int) int {
        if l > r {
            return 0
        }
        if l == r {
            return 1
        }
        key := fmt.Sprintf("%d-%d", l, r)
        if v, ok := cache[key]; ok {
            return v
        }
        if s[l] == s[r] {
            cache[key] = 2 + F(l + 1, r - 1)
        } else {
            cache[key] = max(F(l + 1, r), F(l, r - 1))
        }
        return cache[key]
    }
    return F(0, len(s) - 1)
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}
